Cayley-Dickson (from Rijke-Buchholtz)

Alizter opened the pull request:
This relies on #1150, check the last 2 commits to review.

Here is the Cayley-Dickson construction as described in Rijke-Buchholtz. This gives a H-space on `Join (Sphere 1) (Sphere 1)` when combined with #1151, but we need join associativity to prove that this is `S3`. We could probably prove join associativity directly, there is an "easier" proof with the 3x3 lemma but we don't have this yet.

I'll come back and add some stuff to this later, such as moving some type classes to canonical_names.v and adding the S3 H-space exercise solution to HoTTBookExcersises.

mikeshulman commented:
LGTM.  I also agree with moving those typeclasses into `canonical_names`.

FWIW, I feel like the fact that you can write down the definition of a Cayley-Dickson spheroid like that is a validation of the unbundled approach and the hottclasses design more generally.  Wouldn't it be much more annoying to make that definition with bundled structures?

Alizter commented:
OK, I've rebased and fixed some issues. I am however not finished with this file. There is still some more clean up needing to be done and perhaps a better notation for conjugation needs to be found.

A H-space structure on S3 can be constructed with this but it needs associativity of join.

Alizter commented:
Regarding unbundled structures, I think we should use them as much as possible. But there are some times when they are needed, such as the category of groups.

mikeshulman commented:
Is this PR ready to be reviewed/merged?

Alizter commented:
Not just yet. I would have made this PR a draft if I had known it was possible at the time. I just need to reorganise some stuff and give some lemmas some better names. I haven't had the time to do this yet.

Alizter commented:
@mikeshulman This is ready now, feel free to merge when travis is done.

mikeshulman commented:
Since you just recently changed some stuff, let's wait the 48 hours in case anyone else wants to look at it.
